pkg/kubelet/kubelet_pods.go

```diff
@@ -1020,6 +1022,63 @@ func (kl *Kubelet) removeOrphanedPodStatuses(pods []*v1.Pod, mirrorPods []*v1.Po
 	kl.statusManager.RemoveOrphanedStatuses(podUIDs)
 }
 
+// deleteOrphanedMirrorPods checks orphaned mirror pod whether its grace period has elapsed.
+// Once the grace period passes, podManager.DeleteMirrorPod() is called to delete mirror pod
+// from the API server
+func (kl *Kubelet) deleteOrphanedMirrorPods() {
+	podFullNames := kl.podManager.GetOrphanedMirrorPodNames()
+	start := time.Now()
+	for _, podFullName := range podFullNames {
+		allContainersExited := true
+		m, ok := kl.podManager.GetMirrorPodByFullname(podFullName)
+		if !ok {
+			klog.Warningf("unable to find mirror pod for %q", podFullName)
+			continue
+		}
+		if m.DeletionTimestamp != nil && start.Before(m.DeletionTimestamp.Time) {
+			// within grace period, we would check pod status to see if all its containers have exited
+			name, namespace, err := kubecontainer.ParsePodFullName(podFullName)
+			if err != nil {
+				klog.Warningf("unable to parse fullname %q : %v", podFullName, err)
+				continue
+			}
+			podStatus, err := kl.containerRuntime.GetPodStatus(types.UID(m.Labels[pod.PodUIDLabelKey]), name, namespace)
```

Comment:
just recording:

this logic is moved up to the kubelet and out of pod manager due to checking container runtime for pod status.